### ISSUE TITLE
HasMetadataOperation.periodicWatchUntilReady is broken

### ISSUE THREAD

**Issue by fmichielssen** [fabric8io/kubernetes-client#1515]
Our call to `HasMetadataOperation.waitUntilReady` never returns. A stacktrace suggests that the thread is stuck awaiting the latch at https://github.com/fabric8io/kubernetes-client/blob/5f5d1697839b0b220bf400133a4dbe65a56f797d/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/HasMetadataOperation.java#L192

I believe the cause is https://github.com/fabric8io/kubernetes-client/pull/1451
which changes millisecond values into nanoseconds, but the `interval` value used in the above line is still treated as a milliseconds value, resulting in a huge wait time.

**Comment by oscerd** [comment]
I'm preparing a PR. You're welcome to fix it @fmichielssen if you want

**Comment by paxbit** [comment]
Please reopen. This is still broken and waits for eternity in `io.fabric8.kubernetes.client.internal.readiness.ReadinessWatcher@58` 
because a call like:
`deployableScalableResource.waitUntilReady(30, TimeUnit.SECONDS);` 
leads to this stack:
```
await:58, ReadinessWatcher (io.fabric8.kubernetes.client.internal.readiness)
periodicWatchUntilReady:192, HasMetadataOperation (io.fabric8.kubernetes.client.dsl.base)
waitUntilReady:217, HasMetadataOperation (io.fabric8.kubernetes.client.dsl.base)
waitUntilReady:36, HasMetadataOperation (io.fabric8.kubernetes.client.dsl.base)
```
where it blocks b/c 
`io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation@210` treats the numeric value of `System.nanoTime()` as the time unit the caller set as argument for the `waitUntilReady(...)` call.

Effectively, as in my case I'm using `TimeUnit.SECONDS` for `waitUntilReady(...)`, it does `TimeUnit.SECONDS.toNanos(System.nanoTime())` which yields an incredibly high number definitely unrelated to any intention the method is trying to implement.

**Comment by rohanKanojia** [comment]
@paxbit : apologies for this, will prioritize this for next release

**Comment by oscerd** [comment]
No need to apologize @rohanKanojia .

@paxbit you're welcome to work on a fix

**Comment by rohanKanojia** [comment]
actually, it was me who didn't test this properly. and I updated your PR for that line.

**Comment by rohanKanojia** [comment]
@paxbit @fmichielssen: Polite ping, Is this still an issue? I was trying to reproduce it but could not. Here is the code I used to reproduce it:
```
    try (final KubernetesClient client = new DefaultKubernetesClient()) {
      Pod pod = new PodBuilder()
        .withNewMetadata().withName("myapp-pod").withLabels(Collections.singletonMap("app", "myapp-pod")).endMetadata()
        .withNewSpec()
        .addNewContainer()
        .withName("myapp-container")
        .withImage("busybox:1.28")
        .withCommand("sh", "-c", "echo The app is running!; sleep 15")
        .endContainer()
        .addNewInitContainer()
        .withName("init-myservice")
        .withImage("busybox:1.28")
        .withCommand("sh", "-c", "echo inititalizing...; sleep 5")
        .endInitContainer()
        .endSpec()
        .build();

      String namespace = "default";
      pod = client.pods().inNamespace(namespace).create(pod);
      log("Pod created, waiting for it to get ready");
      client.pods().inNamespace(namespace).withName("myapp-pod").waitUntilReady(30, TimeUnit.SECONDS);
      log("Pod is ready now.");
    }
```

But for me, it returned just after pod got ready. Could you please check if it's working for you on 4.6.1? If still facing problems while waiting, Could you guys please provide a simple test case to reproduce?

I've created a small gif:
![Peek 2019-10-25 14-57](https://user-images.githubusercontent.com/13834498/67560337-4f2ba380-f738-11e9-891f-6c6abfcc8030.gif)

**Comment by rohanKanojia** [comment]
@paxbit @fmichielssen: Guys, did anyone get time to take a look at this again? Shall I close this issue ?

**Comment by paxbit** [comment]
@rohanKanojia Sorry but could you point me to a PR or commit I can actually test b/c I cannot find any changes to HasMetadataOperation after I made my comment on Jul 10th.
https://github.com/fabric8io/kubernetes-client/commits/c6d483f71f599fede155a04465ad8bb5885e6afb/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/HasMetadataOperation.java

Also looking at #1515 and #1530 I fail to see if there was even an attempt to fix it. 

Regarding testing it. If you simply make your POD to not become ready at all then you should see `.waitUntilReady(30, TimeUnit.SECONDS)` to not give up and return after 30 seconds. Instead it will block seemingly indefinitely. While actually, for an input of 30s, it will be waiting for sth. like 922302823974150660 nanoseconds (close to 30years) in: 
`io/fabric8/kubernetes/client/dsl/base/HasMetadataOperation.java:192`

Just set a breakpoint there and try.

**Comment by rohanKanojia** [comment]
@paxbit : You can try testing on `v4.6.3` Does the test case I provided in [this comment](https://github.com/fabric8io/kubernetes-client/issues/1515#issuecomment-546282578) seem valid to you?

Let me try the scenario you described and get back to you.

**Comment by rohanKanojia** [comment]
@paxbit : Hi, I manage to test the scenario you described.

I had a pod whose initcontainer has sleep in it:
```
        try (final KubernetesClient client = new DefaultKubernetesClient()) {
            Pod pod = new PodBuilder()
                    .withNewMetadata().withName("myapp-pod").withLabels(Collections.singletonMap("app", "myapp-pod")).endMetadata()
                    .withNewSpec()
                    .addNewContainer()
                    .withName("myapp-container")
                    .withImage("busybox:1.28")
                    .withCommand("sh", "-c", "echo The app is running!; sleep 10")
                    .endContainer()
                    .addNewInitContainer()
                    .withName("init-myservice")
                    .withImage("busybox:1.28")
                    .withCommand("sh", "-c", "echo initializing...; sleep 36000")
                    .endInitContainer()
                    .endSpec()
                    .build();

            String namespace = "default";
            pod = client.pods().inNamespace(namespace).create(pod);
            log("Pod created, waiting for it to get ready");
            client.resource(pod).inNamespace(namespace).waitUntilReady(30, TimeUnit.SECONDS);
            log("Pod is ready now.");

            LogWatch watch = client.pods().inNamespace(namespace).withName(pod.getMetadata().getName()).watchLog(System.out);
            watch.wait(10);
        }
```
When I ran this code, I was able to get this error:
```
Caused by: io.fabric8.kubernetes.client.KubernetesClientTimeoutException: Timed out waiting for [0] milliseconds for [Pod] with name:[myapp-pod] in namespace [default].
    at io.fabric8.kubernetes.client.internal.readiness.ReadinessWatcher.await (ReadinessWatcher.java:61)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.periodicWatchUntilReady (HasMetadataOperation.java:199)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.periodicWatchUntilReady (HasMetadataOperation.java:208)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.periodicWatchUntilReady (HasMetadataOperation.java:208)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.periodicWatchUntilReady (HasMetadataOperation.java:208)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.periodicWatchUntilReady (HasMetadataOperation.java:208)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.periodicWatchUntilReady (HasMetadataOperation.java:208)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.periodicWatchUntilReady (HasMetadataOperation.java:208)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.periodicWatchUntilReady (HasMetadataOperation.java:208)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.periodicWatchUntilReady (HasMetadataOperation.java:208)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.periodicWatchUntilReady (HasMetadataOperation.java:208)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.periodicWatchUntilReady (HasMetadataOperation.java:208)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.waitUntilReady (HasMetadataOperation.java:224)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.waitUntilReady (HasMetadataOperation.java:36)
    at io.fabric8.kubernetes.client.handlers.PodHandler.waitUntilReady (PodHandler.java:88)
    at io.fabric8.kubernetes.client.handlers.PodHandler.waitUntilReady (PodHandler.java:39)
    at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.waitUntilReady (NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.java:246)
    at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.waitUntilReady (NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.java:59)
    at io.fabric8.WaitUntilReadyDemo.main (WaitUntilReadyDemo.java:42)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:293)
    at java.lang.Thread.run (Thread.java:834)
[WARNING] thread Thread[OkHttp ConnectionPool,5,io.fabric8.WaitUntilReadyDemo] was interrupted but is still alive after waiting at least 15000msecs
[WARNING] thread Thread[OkHttp ConnectionPool,5,io.fabric8.WaitUntilReadyDemo] will linger despite being asked to die via interruption
[WARNING] thread Thread[Okio Watchdog,5,io.fabric8.WaitUntilReadyDemo] will linger despite being asked to die via interruption
[WARNING] thread Thread[OkHttp Http2Connection,5,io.fabric8.WaitUntilReadyDemo] will linger despite being asked to die via interruption
[WARNING] NOTE: 3 thread(s) did not finish despite being asked to  via interruption. This is not a problem with exec:java, it is a problem with the running code. Although not serious, it should be remedied.
[WARNING] Couldn't destroy threadgroup org.codehaus.mojo.exec.ExecJavaMojo$IsolatedThreadGroup[name=io.fabric8.WaitUntilReadyDemo,maxpri=10]
java.lang.IllegalThreadStateException
    at java.lang.ThreadGroup.destroy (ThreadGroup.java:776)
    at org.codehaus.mojo.exec.ExecJavaMojo.execute (ExecJavaMojo.java:328)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  47.189 s
[INFO] Finished at: 2019-11-19T21:45:10+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.4.0:java (default-cli) on project kubernetes-client-demo: An exception occured while executing the Java class. null: InvocationTargetException: Timed out waiting for [0] milliseconds for [Pod] with name:[myapp-pod] in namespace [default]. -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
~/work/repos/kubernetes-client-demo : $ 
```
![Peek 2019-11-19 21-45](https://user-images.githubusercontent.com/13834498/69164825-b4479e80-0b16-11ea-8699-ae14da3aeb63.gif)

**Comment by rohanKanojia** [comment]
I did all this testing on v4.6.3

**Comment by paxbit** [comment]
@rohanKanojia Your scenario seemed valid to me. LGTM now when testing with 4.6.3. You can close if you like. 

BTW: I would have more easily noticed 450b94745b68403293a55956be2aa7ec483c0a6c, which contains the actual fix, if it were in any way related to #1515 or #1530. It would have been helpful to update either of those two with a reference to that commit after pushing it. 

Thank you for your support!

**Comment by rohanKanojia** [comment]
@paxbit : Thanks a lot for taking time to test this. I apologize for not keeping track of this issue while checking PRs.

Closing this one.
